src/query/server/query.go

```diff
@@ -484,6 +485,29 @@ func Run(runOpts RunOptions) RunResult {
 		if err != nil {
 			logger.Fatal("unable to setup downsampler for m3db backend", zap.Error(err))
 		}
+	case config.PromRemoteStorageType:
+		var opts promremote.Options
+		opts, err = promremote.NewOptions(cfg.PrometheusRemoteBackend, scope)
+		if err != nil {
+			logger.Fatal("invalid configuration", zap.Error(err))
+		}
+		backendStorage, err = promremote.NewStorage(opts)
+		if err != nil {
+			logger.Fatal("unable to setup prom remote backend", zap.Error(err))
+		}
+		defer func() {
+			err := backendStorage.Close()
+			if err != nil {
+				logger.Error("error when closing storage", zap.Error(err))
+			}
+		}()
+
+		downsampler, clusterClient, err = newDownsamplerAsync(cfg.Downsample, cfg.ClusterManagement.Etcd, backendStorage,
+			clusterNamespacesWatcher, tsdbOpts.TagOptions(), clockOpts, instrumentOptions, rwOpts, runOpts,
+		)
```

Comment:
Not sure if this breaks anything, but I believe `clusterNamespacesWatcher` here is not initialized. At least m3db `case` calls `initClusters()`

https://github.com/m3db/m3/blob/43a0942cf449892eddc04aa004ab63bd7ede3f3c/src/query/server/query.go#L454-L458

Also, what's unusual is that the cleanup of `clusterNamespaceWatcher` is defined in `newM3DBStorage()` function and not where it's initialized (ditto for `clusters` in the snippet below). It might make sense moving the cleanup out of there.

https://github.com/m3db/m3/blob/43a0942cf449892eddc04aa004ab63bd7ede3f3c/src/query/server/query.go#L720-L736

Reply:
Yeah this one is confusing. I will need to double check how this namespace watcher is used in downsampler and writer. I might need to register namespaces to it from endpoints. I think its ok if its not initialised since some backends (e.g grpc backend) do not initialise it. But I will move it out of `newM3DBStorage` to avoid similar mistake in the future.